### PR TITLE
feat: hard logout option [AR-1653]

### PR DESCRIPTION
**Pull request by saleniuk**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-1653" title="AR-1653" target="_blank">StoryAR-1653</a>  Wipe History Option on Log-out
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We need to have the option to choose whether we want to do a hard or soft logout.

### Solutions

Added new dialog with a checkbox when clicking on "logout" to choose, navigation after logout moved to `SelfUserProfileViewModel`, used new use case `getOrRegisterClient` when logging in.

### Dependencies (Optional)

Needs releases with:

- https://github.com/wireapp/kalium/pull/880

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

Open self profile, click on logout and choose to delete all personal data or not.

### Attachments (Optional)

https://user-images.githubusercontent.com/30429749/189762785-608096cf-bf93-41ef-b45e-b9ca8c920eb5.mp4

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-970/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-970/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4405343)(4405343)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4405343)(4405343)-dev-release.apk)

File: app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileScreen.kt [hunk 1]
```diff
@@ -115,12 +115,15 @@ private fun SelfUserProfileContent(
         }
     }
     val scrollState = rememberScrollState()
+    val logoutWipeDataDialogState = rememberVisibilityState<LogoutWipeDataDialogState>()
```
Comment (mchenani) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileScreen.kt]: What do you think of renaming the dialog to LogoutOptionsDialog? :D and here logoutOptionsDailogState?

File: app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt [hunk 2]
```diff
@@ -159,10 +160,21 @@ class SelfUserProfileViewModel @Inject constructor(
...
-    fun onLogoutClick() {
+    fun logout(wipeData: Boolean) {
         viewModelScope.launch {
-            logout()
-            dataStore.clear() // TODO this should be moved to some service that will clear all the data in the app
+            val selfUserId = getSelf().first().id
+            val otherUserId = userProfileState.otherAccounts.firstOrNull()?.id
+            logout(reason = LogoutReason.SELF_LOGOUT, isHardLogout = wipeData)
+            if (wipeData) {
+                dataStore.clear() // TODO this should be moved to some service that will clear all the data in the app
```
Comment (mchenani) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt:167]: Wasn't this part of the logout useCase?

File: app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt [hunk 3]
```diff
@@ -159,10 +160,21 @@ class SelfUserProfileViewModel @Inject constructor(
...
+            }
+            otherUserId?.let {
+                switchAccount(it)
```
Comment (mchenani) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt]: In the last code whenever something happens to the session in one place we were managing switches and etc. but here we have the logic in multiple places, I think this part is mixing stuff! if you like we can discuss it in 1:1.

**Review by mchenani — CHANGES_REQUESTED** [review]
I have some question regarding mixing logic in the view model, I think the previous logic was better.

File: app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileScreen.kt (hunk 1 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileScreen.kt]: Nice idea, I changed the name 👍

File: app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt (hunk 2 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt:167]: I kept it the way it was, this `dataStore` is the AR thing only, so we can't move it to the kalium's `LogoutUseCase`. There is a TODO to not forget about it and clear it up.

File: app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt (hunk 3 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/self/SelfUserProfileViewModel.kt]: My bad, I reverted it.

**Review by mchenani — APPROVED** [review]
🚀

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-970/5/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4409346)(4409346)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4409346)(4409346)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-970/7/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4409658)(4409658)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4409658)(4409658)-dev-release.apk)

**Pull request merged**